### PR TITLE
MMA 868 - Move se-mailbox to Python 3

### PR DESCRIPTION
**Pull request by denisa-bortanoiu**
https://jira.solarwinds.com/browse/MMA-868

### CODE REVIEW

File: se_mailbox/se_mailbox.py [hunk 1]
```diff
@@ -145,9 +145,18 @@ def size(self):
         total_size = 0
         total_count = 0
         for i, line in enumerate(open(self.size_fn).readlines()[1:]):
-            size, count = line.split()
-            total_size += int(size)
-            total_count += int(count)
+            if not line or len(line.split()) != 2:
```
Comment (cosmingrz) [inline se_mailbox/se_mailbox.py]: I don't understand why we changed these lines.

File: se_mailbox/smaildir.py [hunk 2]
```diff
@@ -23,7 +24,13 @@ class Maildir(mailbox.Maildir):
 
     def list_folders(self):
         """Return a list of folder names."""
-        return list(self.iter_folders)
```
Comment (cosmingrz) [inline se_mailbox/smaildir.py]: Why we changed this and we added duplicate code?

File: se_mailbox/smaildir.py [hunk 3]
```diff
@@ -36,19 +43,17 @@ def iter_folders(self):
...
-        def raise_if_not_empty(subpath):
-            """Raise a NotEmptyError if the folder is not empty."""
-            for entry in scandir.scandir(os.path.join(path, subpath)):
-                if len(entry.name) < 1 or entry.name[0] != '.':
-                    raise mailbox.NotEmptyError(
-                        'Folder contains message(s): %s' % folder)
-
-        raise_if_not_empty("new")
-        raise_if_not_empty("cur")
+        for entry in scandir.scandir(os.path.join(path, 'new')):
```
Comment (cosmingrz) [inline se_mailbox/smaildir.py]: This was changed in order to make unit testing easy?

File: setup.py [hunk 4]
```diff
@@ -38,12 +39,11 @@
         'Development Status :: 2 - Pre-Alpha',
         'Intended Audience :: Developers',
         'Natural Language :: English',
-        "Programming Language :: Python :: 2",
-        'Programming Language :: Python :: 2.7',
         'Programming Language :: Python :: 3',
```
Comment (cosmingrz) [inline setup.py]: I guess we want to support only python 3.7. In this case we should remove these lines.

File: se_mailbox/filelock.py [hunk 5]
```diff
@@ -34,16 +34,6 @@ def __init__(self, file_name, timeout=10, delay=0.05, base=None,
         self.timeout = timeout
         self.delay = delay
         self.fd = None
-        if max_age and os.path.exists(self.lockfile):
```
Comment (Silviu-Surcica) [inline se_mailbox/filelock.py]: Why is this removed?

File: se_mailbox/smaildir.py [hunk 6]
```diff
@@ -36,19 +43,17 @@ def iter_folders(self):
...
+                raise mailbox.NotEmptyError('Folder contains message(s): %s' %
+                                            folder)
+        for entry in scandir.scandir(os.path.join(path, 'cur')):
+            if len(entry.name) < 1 or entry.name[0] != '.':
+                raise mailbox.NotEmptyError('Folder contains message(s): %s' %
+                                            folder)
         for entry in scandir.scandir(path):
-            if (entry.name != 'new' and entry.name != 'cur' and
-                   entry.name != 'tmp' and entry.is_dir()):
+            if entry.name != 'new' and entry.name != 'cur' and \
```
Comment (Silviu-Surcica) [inline se_mailbox/smaildir.py]: why is this change needed ? PEP8 encourages to use parentheses for multi line code.

File: se_mailbox/smaildir.py (hunk 3 above)
Comment (Silviu-Surcica) [reply se_mailbox/smaildir.py]: why the need to duplicate code?

File: se_mailbox/se_mailbox.py (hunk 1 above)
Comment (RoxanaButuc) [reply se_mailbox/se_mailbox.py]: Seems that we want to make sure that each processed line is in the correct format (having 2 values, no extra spaces and the correct type) in order to avoid errors.

File: README.rst [hunk 7]
```diff
@@ -38,7 +32,14 @@ Additional mailbox functionality.
 Features
 --------
 
-* TODO
+* ``filelock.Filelock`` offers a file locking mechanism that has context-manager support so
+    you can use it in a with statement.
+
+* ``smaildir.Maildir`` is a Gmail-style Maildir mailbox. It uses scandir() instead of listdir(). It also provides a ew iter_folders() method that works like list_folders(), put provides a generator instead of returning a list.
```
Comment (tonyandrewmeyer) [inline README.rst]: There's nothing "gmail-style" about this?

File: README.rst [hunk 8]
```diff
@@ -38,7 +32,14 @@ Additional mailbox functionality.
...
+
+* ``se_mailbox.QuotaMixin`` implements the Maildir++ quota size system, as described here: http://www.inter7.com/courierimap/README.maildirquota.html. Quotas are not enforced - this would be good to add, but since we do a non-standard operation (dropping old data instead of failing to add)    there's not much point.
```
Comment (tonyandrewmeyer) [inline README.rst]: The odd extra spaces should be removed.

File: se_mailbox/filelock.py [hunk 9]
```diff
@@ -19,11 +20,10 @@ class FileLockException(Exception):
 
 class FileLock(object):
     """A file locking mechanism that has context-manager support so
-    you can use it in a with statement. This should be relatively cross
-    compatible as it doesn't rely on msvcrt or fcntl for the locking."""
+    you can use it in a with statement.
```
Comment (tonyandrewmeyer) [inline se_mailbox/filelock.py]: Why did we change this so it uses fcntl?

File: se_mailbox/smaildir.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
 """Read/write support for Maildir mailboxes."""
 
+
```
Comment (tonyandrewmeyer) [inline se_mailbox/smaildir.py]: Why was this blank line added?

File: se_mailbox/se_mailbox.py (hunk 1 above)
Comment (tonyandrewmeyer) [reply se_mailbox/se_mailbox.py]: That doesn't seem like a Python 2->3 change?

**Comment by denisa-bortanoiu** [comment]
The code changes were done because we required to get the package up to date with mail-core. As far as I remember, I updated the code so that it's the same as it currently is mail-core.

File: README.rst (hunk 7 above)
Comment (RoxanaButuc) [reply README.rst]: Done in 3956849b72c52f9ec08cc9240a89e4d3eaad5224

File: README.rst [hunk 11]
```diff
@@ -38,7 +32,14 @@ Additional mailbox functionality.
 Features
 --------
 
-* TODO
+* ``filelock.Filelock`` offers a file locking mechanism that has context-manager support so
+    you can use it in a with statement.
+
+* ``smaildir.Maildir`` offers support for maildir mailbox. It uses scandir() instead of listdir(). It also provides a new iter_folders() method that works like list_folders(), put provides a generator instead of returning a list.
+
+* ``se_mailbox.QuotaMixin`` implements the Maildir++ quota size system, as described here: http://www.inter7.com/courierimap/README.maildirquota.html. Quotas are not enforced - this would be good to add, but since we do a non-standard operation (dropping old data instead of failing to add) there's not much point.
```
Comment (tonyandrewmeyer) [inline README.rst]: I think the "since we do a non-standard operation" part of this should probably not be there on a public repo.

**Comment by RoxanaButuc** [comment]
[Remove refactoring changes](https://github.com/SpamExperts/se-mailbox/pull/170/commits/59386bc36f7f59d9ebbe8990cd92d678a1ce8bc4)

File: se_mailbox/filelock.py (hunk 9 above)
Comment (tonyandrewmeyer) [reply se_mailbox/filelock.py]: @RoxanaButuc unanswered query here.

File: se_mailbox/filelock.py (hunk 5 above)
Comment (tonyandrewmeyer) [reply se_mailbox/filelock.py]: @RoxanaButuc unanswered query here.

File: se_mailbox/smaildir.py (hunk 3 above)
Comment (tonyandrewmeyer) [reply se_mailbox/smaildir.py]: @RoxanaButuc unanswered query here.

File: se_mailbox/filelock.py (hunk 9 above)
Comment (RoxanaButuc) [reply se_mailbox/filelock.py]: Changes were made based on this [comment](https://jira.solarwinds.com/browse/MMA-868?focusedCommentId=1714710&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1714710) - in order to get the package up to date with mail-core

File: se_mailbox/filelock.py (hunk 5 above)
Comment (RoxanaButuc) [reply se_mailbox/filelock.py]: Changes were made based on this [comment](https://jira.solarwinds.com/browse/MMA-868?focusedCommentId=1714710&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1714710) - in order to get the package up to date with mail-core

File: se_mailbox/smaildir.py (hunk 3 above)
Comment (RoxanaButuc) [reply se_mailbox/smaildir.py]: This was change in order to get the package up to date with mail-core (smaildir from se-mailboxes and the one from mail-core).
